File: libs/mer_graph/edgemer.py
import argparse

def read_seq(file):
    """Reads the sequence file and yields edges one at a time."""
    with open(file, 'r') as f:
        for line in f:
            parts = line.split()
            parts.pop(0)  # remove name
            nlist = [j[:len(j)-1] for j in parts]
            slist = [j[len(j)-1:] for j in parts]
            pcount = 0
            for k in range(len(nlist) - 1):
                if pcount >= 0:  # vote is for primary strand
                    yield [nlist[k] + slist[k], nlist[k+1] + slist[k+1]]
                else:
                    yield [nlist[k+1] + ('+' if slist[k+1] == '-' else '-'), 
                           nlist[k] + ('+' if slist[k] == '-' else '-')]

def revcomp(seq):
    """Computes the reverse complement of a DNA sequence."""
    complement = str.maketrans("ACGT", "TGCA")
    return seq.translate(complement)[::-1]

def read_seg(file, edges, K):
    """Reads the segment file and yields edge-mer pairs one at a time."""
    nodeSeg = {}
    with open(file, 'r') as f:
        for line in f:
            l = line.split()
            nodeSeg[l[0] + '+'] = l[1]
            nodeSeg[l[0] + '-'] = revcomp(l[1])
    
    for edge in edges:
        mer = nodeSeg[edge[0]][-K] + nodeSeg[edge[1]][:K]
        yield edge[0] + ' ' + edge[1], mer

def write_graph(file, edges_func, edgeMers):
    """Writes the graph to the output file, processing data on-the-fly."""
    nodes = set()
    with open(file, 'w') as f:
        f.write('#graph 1\n')
        # First pass: count unique nodes
        for edge in edges_func():  # Call the generator function again
            nodes.add(edge[0])
            nodes.add(edge[1])
        f.write(str(len(nodes)) + '\n')
        # Second pass: write edge-mer pairs
        for e, m in edgeMers:
            f.write(e + ' ' + m + '\n')

def main():
    parser = argparse.ArgumentParser(prog='edgecnt',
                                     description='creates a count graph based on cuttlefish output and fasta')
    parser.add_argument('-k', '--kvalue', required=True, type=int, help='k-mer size')
    parser.add_argument('-c', '--cuttlefishprefix', required=True, help='prefix for cuttlefish output files')
    parser.add_argument('-o', '--output', required=True, help='output graph file')
    args = parser.parse_args()

    print("reading seq")
    edges_func = lambda: read_seq(args.cuttlefishprefix + '.cf_seq')  # Store generator function
    edges = edges_func()  # Create generator for the first iteration
    print("reading seg")
    edgeMers = read_seg(args.cuttlefishprefix + '.cf_seg', edges, args.kvalue)  # Generator for edge-mer pairs
    print("writing graph")
    write_graph(args.output, edges_func, edgeMers)  # Pass the generator function

if __name__ == '__main__':
    main()

File: libs/mer_graph/src/edgemer.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::path::Path;

// Work in progress, sorry //
fn read_seq(file: &str) -> Vec<[String; 2]> {
    let mut edges = Vec::new();
    let lines = read_lines(file).expect("Could not read lines");

    for line in lines {
        if let Ok(line) = line {
            let mut parts: Vec<_> = line.split_whitespace().collect();
            parts.remove(0); // Remove the name
            let (mut nlist, mut slist): (Vec<String>, Vec<String>) = parts
                .iter()
                .map(|&j| {
                    let len = j.len();
                    (j[..len - 1].to_string(), j[len - 1..].to_string())
                })
                .unzip();

            for k in 0..nlist.len() - 1 {
                let edge = if k >= 0 {
                    [nlist[k].clone() + &slist[k], nlist[k + 1].clone() + &slist[k + 1]]
                } else {
                    [
                        nlist[k + 1].clone() + if slist[k + 1] == "-" { "+" } else { "-" },
                        nlist[k].clone() + if slist[k] == "-" { "+" } else { "-" },
                    ]
                };
                edges.push(edge);
            }
        }
    }
    edges
}

fn revcomp(seq: &str) -> String {
    let mut rcseq = seq
        .replace("A", "t")
        .replace("C", "g")
        .replace("T", "a")
        .replace("G", "c")
        .to_uppercase();
    rcseq = rcseq.chars().rev().collect::<String>();
    rcseq
}

fn read_seg(file: &str, edges: Vec<[String; 2]>, k: usize) -> HashMap<String, String> {
    let mut node_seg = HashMap::new();
    if let Ok(lines) = read_lines(file) {
        for line in lines {
            if let Ok(line) = line {
                let parts: Vec<_> = line.split_whitespace().collect();
                node_seg.insert(parts[0].to_string() + "+", parts[1].to_string());
                node_seg.insert(parts[0].to_string() + "-", revcomp(parts[1]));
            }
        }
    }

    let mut edge_mers = HashMap::new();
    for edge in edges {
        let mer = format!(
            "{}{}",
            &node_seg[&edge[0]][node_seg[&edge[0]].len() - k..],
            &node_seg[&edge[1]][..k]
        );
        edge_mers.insert(edge.join(" "), mer);
    }

    edge_mers
}

fn write_graph(file: &str, edges: Vec<[String; 2]>, edge_mers: HashMap<String, String>) {
    let mut nodes = HashSet::new();
    for e in edges.iter() {
        nodes.insert(&e[0]);
        nodes.insert(&e[1]);
    }

    let mut output = File::create(file).expect("Could not create output file");
    writeln!(output, "#graph 1").expect("Failed to write");
    writeln!(output, "{}", nodes.len()).expect("Failed to write");

    for (e, m) in edge_mers.iter() {
        writeln!(output, "{} {}", e, m).expect("Failed to write");
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!("Usage: {} -k <kvalue> -c <cuttlefishprefix> -o <output>", args[0]);
        return;
    }

    let kvalue: usize = args[2].parse().expect("Invalid K value");
    let cuttlefish_prefix = &args[4];
    let output_file = &args[6];

    println!("reading seq");
    let edges = read_seq(&(cuttlefish_prefix.clone() + ".cf_seq"));
    println!("reading seg");
    let edge_mers = read_seg(&(cuttlefish_prefix.clone() + ".cf_seg"), edges, kvalue);
    println!("writing graph");
    write_graph(output_file, edges, edge_mers);
}
